openedx_lti_tool_plugin/deep_linking/forms.py

```diff
@@ -1,59 +1,71 @@
 """Django Forms."""
+import json
 from typing import List, Optional, Set, Tuple
 
 from django import forms
 from django.http.request import HttpRequest
 from django.urls import reverse
 from django.utils.translation import gettext as _
+from pylti1p3.contrib.django.lti1p3_tool_config.models import LtiTool
 from pylti1p3.deep_link_resource import DeepLinkResource
 
 from openedx_lti_tool_plugin.apps import OpenEdxLtiToolPluginConfig as app_config
+from openedx_lti_tool_plugin.deep_linking.exceptions import DeepLinkingException
 from openedx_lti_tool_plugin.edxapp_wrapper.learning_sequences import course_context
+from openedx_lti_tool_plugin.models import CourseAccessConfiguration
+from openedx_lti_tool_plugin.waffle import COURSE_ACCESS_CONFIGURATION
 
 
 class DeepLinkingForm(forms.Form):
     """Deep Linking Form."""
 
-    def __init__(self, *args: tuple, request=None, **kwargs: dict):
+    content_items = forms.MultipleChoiceField(
+        required=False,
+        widget=forms.CheckboxSelectMultiple,
+        label=_('Courses'),
+    )
+
+    def __init__(
+        self,
+        *args: tuple,
+        request: HttpRequest,
+        lti_tool: LtiTool,
+        **kwargs: dict,
+    ):
         """Class __init__ method.
 
-        Initialize class instance attributes and add `content_items` field.
+        Initialize class instance attributes and add field choices
+        to the `content_items` field.
 
         Args:
             *args: Variable length argument list.
             request: HttpRequest object.
+            lti_tool: LtiTool model instance.
             **kwargs: Arbitrary keyword arguments.
 
         """
         super().__init__(*args, **kwargs)
-        self.fields['content_items'] = forms.MultipleChoiceField(
-            choices=self.get_content_items_choices(request),
-            required=False,
-            widget=forms.CheckboxSelectMultiple,
-            label=_('Courses'),
-        )
+        self.request = request
+        self.lti_tool = lti_tool
+        self.fields['content_items'].choices = self.get_content_items_choices()
 
-    def get_content_items_choices(self, request: HttpRequest) -> List[Tuple[str, str]]:
+    def get_content_items_choices(self) -> List[Tuple[str, str]]:
         """Get `content_items` field choices.
 
-        Args:
-            request: HttpRequest object.
-
         Returns:
             List of tuples with choices for the `content_items` field.
 
         """
         return [
-            self.get_content_items_choice(course, request)
-            for course in course_context().objects.all()
+            self.get_content_items_choice(course)
```

Comment:
@kuipumu Same question of the last PR review, Is it necessary that the return of the function be a list? I mention it because of the performance where for example, searching through 100,000 items takes 49.663 seconds with a list, but only 0.007 seconds with a set, taking into account that there are no repeated values in this array.

Reply:
@anfbermudezme I didn't tried with a set since the Django documentation specifies the choices field to be a list of tuples, however it works with a set so I change this.

Reply:
I think that is better to compare using a set vs a list, since iterating over a list would be faster than in a set, I'm not sure if Django performs more iteration vs lookup for this particular field, but it seems that for validating the choices it iterates over all the choices (https://github.com/django/django/blob/stable/3.2.x/django/forms/fields.py#L779), if it happens to be a set, then it will have a higher complexity than using a list. Maybe the best approach to decide here would be to test the initialization time for using a set vs a list. What do you think @anfbermudezme @kuipumu ?